Lesson19/DynamicIntArray.cpp

```diff
@@ -2,53 +2,63 @@
 #include <iostream>
 
 // Default c-tor
-DynamicIntArray::DynamicIntArray() :m_size(5), m_position(0) {
+DynamicIntArray::DynamicIntArray() :m_size(0), m_position(0) {
     m_elements = new int[m_size];
+    std::cout << "Defalult c-tor" << std::endl;
 }
 
 // C-tor with size
 DynamicIntArray::DynamicIntArray(std::size_t size) : m_size(size), m_position(0) {
     m_elements = new int[m_size];
+    std::cout << "C-tor with size" << std::endl;
 }
 
 // Copy c-tor
 DynamicIntArray::DynamicIntArray(const DynamicIntArray& other) {
-    if (m_size < other.getSize()) {
-        delete[] m_elements;
-        m_size = other.m_size;
-    }
-
-    m_elements = new int[m_size];
-    for (int i = 0; i < m_size; i++) {
-        m_elements[i] = other.m_elements[i];
-    }
+    m_size = other.m_size;
     m_position = other.m_position;
+    m_elements = new int[other.m_size];
+
+    std::copy(other.m_elements, other.m_elements + other.m_size, m_elements);
+    std::cout << "Copy c-tor" << std::endl;
 }
 
 // Free allocated memory
 DynamicIntArray::~DynamicIntArray() {
+    std::cout << "D-tor" << std::endl;
     delete[] m_elements;
 }
 
 DynamicIntArray& DynamicIntArray::operator=(const DynamicIntArray& other) {
-    return *this;
+    DynamicIntArray tmp(other);
+    delete[] m_elements;
+    return tmp;
 }
 
 int& DynamicIntArray::operator[](std::size_t index) {
+    if (index >= m_size) {
+        index = m_size - 1;
+    }
```

Comment:
it's wrong behaviour. if index is out of range you should throw an exception or use assert

Reply:
added assert